python/hail/representation/pedigree.py

```diff
@@ -0,0 +1,305 @@
+
+        The considered fields are ``mother``, ``father``, and ``sex``.
+        Recall that ``proband`` may not ever be missing. The ``fam`` field 
+        may be missing in a complete trio.
+
+        :rtype: bool
+        """
+
+        if not hasattr(self, '_complete'):
+            self._complete = self._jrep.isComplete()
+        return self._complete
+
+
+class Pedigree(object):
+    """Class containing a list of trios, with extra functionality.
+
+    :param trios: list of trio objects to include in pedigree
+    :type trios: list of :class:`.Trio`
+    """
+
+    @handle_py4j
+    def __init__(self, trios):
+        if not isinstance(trios, list):
+            raise TypeError("parameter 'trios' must be of type list, but found '%s'" % type(trios))
+        for t in trios:
+            if not isinstance(t, Trio):
+                raise TypeError("all elements of list 'trios' must be of type Trio, but found '%s'" % type(t))
+
+        self._jrep = Env.hail().methods.Pedigree(jindexed_seq([t._jrep for t in trios]))
+        self._trios = trios
+        self._complete_trios = None
+
+    @classmethod
+    def _from_java(cls, jrep):
+        ped = Pedigree.__new__(cls)
+        ped._jrep = jrep
+        ped._trios = None
+        ped._complete_trios = None
+        return ped
+
+    def __eq__(self, other):
+        if not isinstance(other, Pedigree):
+            return False
+        else:
+            return self._jrep == other._jrep
+
+    def __hash__(self):
+        return self._jrep.hashCode()
+
+    @staticmethod
+    @handle_py4j
+    def read_fam(fam_path, delimiter='\\s+'):
+        """Read a .fam file and return a pedigree object.
+        
+        **Examples**
+        
+        >>> ped = Pedigree.read_fam('data/test.fam')
+        
+        **Notes**
+
+        This method reads a `PLINK .fam file <https://www.cog-genomics.org/plink2/formats#fam>`_.
+
+        Hail expects a file in the same spec as PLINK outlines.
+
+        :param str fam_path: path to .fam file.
+        
+        :param str delimiter: Field delimiter.
+
+        :rtype: :class:`.Pedigree`
+        """
+
+        jrep = scala_object(Env.hail().methods, 'Pedigree').fromFam(fam_path, Env.hc()._jhc.hadoopConf(), delimiter)
+        return Pedigree._from_java(jrep)
+
+    @property
+    def trios(self):
+        """List of trio objects in this pedigree.
+
+        :rtype: list of :class:`.Trio`
+        """
+
+        if not self._trios:
+            self._trios = [Trio._from_java(t) for t in jiterable_to_list(self._jrep.trios())]
+        return self._trios
+
+    @property
+    def complete_trios(self):
+        """List of trio objects that have a defined father, mother, and sex.
+
+        :rtype: list of :class:`.Trio`
+        """
+
+        if not self._complete_trios:
+            self._complete_trios = filter(lambda t: t.is_complete(), self._trios)
+        return self._complete_trios
```

Comment:
This is a function down below.  Should it just be function here, too?  That is, no caching?